NERC-CEH/plankton_ml
# Proof of concept of similarity search with the scivision model

This builds up the demo to show a reasonable proof of concept, for a human viewer, of enough success in extracting image embeddings from the `scivision` plankton model to keep building experiments on it in the short term.

## What's in this

* Adds some utility functions to make working with the image embeddings simpler
* Changes the way the `intake` catalogue is written to use the whole untagged image collection, not the subset of labelled ones
* Adds a bit more test coverage where it was spotty, moves the tests into the package for the `flake8` action to collect them
* Adapts the `image_embeddings.py` script to run through the whole collection through the model
* Adds a [notebook showing the outcome of similarity search](https://github.com/NERC-CEH/plankton_ml/blob/more_data/notebooks/VectorSearch.ipynb) of the embeddings, which looks plausible, and some notes on next steps / link to a related paper working through unsupervised clustering approaches 

## What isn't in this

Any kind of robust approach to spreading the workload around with Dask, partly the collection isn't big enough to justify it yet, partly i haven't worked with dask enough yet to understand the best approach, and would appreciate any advice on that topic, including from the DevOps folks (see the notes in comments in `scripts/image_embeddings.py`). 

Any exploration of [model explainability techniques](https://captum.ai/) using the prediction capabilities of the CEFAS model as opposed to using it as a source of embeddings. That's a good place to visit next, before trying any [clustering algorithms on the embeddings](https://sslneurips23.github.io/paper_pdfs/paper_70.pdf)- step back and observe whether what the model is seeing is properly coherent, whether it aligns with the CEFAS reference data, or whether there are factors like image dimensions giving a false positive impression of these initial results. 

![A plot of the closest 24 samples to one picked at random](https://github.com/NERC-CEH/plankton_ml/assets/369631/4e6790cb-1bc0-421a-a2a3-069040f63c1b)
 
## To test

`export PYTHONPATH=.` or `pip install -e .`
`py.test`

You may need the right credentials in `.env` to be able to run the scripts which generate the index and subsequently the embeddings. I've only run this locally, there are 8k images. I should set it up to be able to reproduce this using only the images from @Kzra that are in the test fixtures.
